sky/data/storage_utils.py

```diff
@@ -120,91 +123,99 @@ def get_excluded_files_from_gitignore(src_dir_path: str) -> List[str]:
+
+        if (e.returncode == exceptions.GIT_FATAL_EXIT_CODE and
+                'not a git repository' in e.stderr):
+            # If git failed because we aren't in a git repository, but there is
+            # a .gitignore, warn the user that it will be ignored.
+            if os.path.exists(gitignore_path):
+                logger.warning('Detected a .gitignore file, but '
+                               f'{src_dir_path} is not a git repository. The '
+                               '.gitignore file will be ignored. '
+                               f'{_USE_SKYIGNORE_HINT}')
+            # Otherwise, this is fine and we can exit early.
+            return []
+
+        if e.returncode == exceptions.COMMAND_NOT_FOUND_EXIT_CODE:
+            # Git is not installed. This is fine, skip the check.
+            # If .gitignore is present, warn the user.
+            if os.path.exists(gitignore_path):
+                logger.warning(f'Detected a .gitignore file in {src_dir_path}, '
+                               'but git is not installed. The .gitignore file '
+                               f'will be ignored. {_USE_SKYIGNORE_HINT}')
+            return []
+
+        # Pretty much any other error is unexpected, so re-raise.
+        raise
 
-    git_exclude_exists = os.path.isfile(git_exclude_path)
-    gitignore_exists = os.path.isfile(gitignore_path)
+    # submodules_output will contain each submodule path (relative to
+    # src_dir_path), each ending with a null character.
+    # .split will have an empty string at the end because of the final null
+    # char, so trim it.
+    submodules = submodules_output.stdout.split('\0')[:-1]
+
+    # The empty string is the relative reference to the src_dir_path.
+    all_git_repos = ['.'] + [
+        # We only care about submodules that are a subdirectory of src_dir_path.
+        submodule for submodule in submodules if not submodule.startswith('../')
+    ]
 
-    # This command outputs a list to be excluded according to .gitignore
-    # and .git/info/exclude
-    filter_cmd = (f'git -C {shlex.quote(expand_src_dir_path)} '
-                  'status --ignored --porcelain=v1')
     excluded_list: List[str] = []
+    for repo in all_git_repos:
+        # repo is the path relative to src_dir_path. Get the full path.
+        repo_path = os.path.join(expand_src_dir_path, repo)
+        # This command outputs a list to be excluded according to .gitignore
+        # and .git/info/exclude
+        filter_cmd = (f'git -C {shlex.quote(repo_path)} ls-files -z '
+                      '--others --ignore --exclude-standard --directory')
+        output = subprocess.run(filter_cmd,
+                                shell=True,
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE,
+                                check=True,
+                                text=True)
+        # Don't catch any errors. We would only expect to see errors during the
```

Comment:
Similar as above, wondering if we should optimize for the case that `jobs launch` should not crash if any such issue happens. Not feeling strongly.

Reply:
same as above